swarm/swap/protocol.go

```diff
@@ -0,0 +1,200 @@
+		swap:  swapAccount,
+	}
+	return proto
+}
+
+// NewPeer is the constructor for Peer
+func NewPeer(peer *protocols.Peer, swap *SwapProtocol) *SwapProtocolPeer {
+	p := &SwapProtocolPeer{
+		Peer:         peer,
+		swapProtocol: swap,
+	}
+	return p
+}
+
+type IssueChequeMsg struct {
+	Cheque *Cheque
+}
+
+type RedeemChequeMsg struct {
+}
+
+/////////////////////////////////////////////////////////////////////
+// SECTION: node.Service interface
+/////////////////////////////////////////////////////////////////////
+
+func (p *SwapProtocol) Start(srv *p2p.Server) error {
+	log.Debug("Started swap")
+	return nil
+}
+
+func (p *SwapProtocol) Stop() error {
+	log.Info("swap shutting down")
+	return nil
+}
+
+var swapSpec = &protocols.Spec{
+	Name:       swapProtocolName,
+	Version:    swapVersion,
+	MaxMsgSize: defaultMaxMsgSize,
+	Messages: []interface{}{
+		SwapMsg{},
+	},
+}
+
+func (p *SwapProtocol) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    swapSpec.Name,
+			Version: swapSpec.Version,
+			Length:  swapSpec.Length(),
+			Run:     p.Run,
+		},
+	}
+}
+
+func (p *SwapProtocol) APIs() []rpc.API {
+	apis := []rpc.API{
+		{
+			Namespace: "swap",
+			Version:   "1.0",
+			Service:   NewAPI(p),
+			Public:    true,
+		},
+	}
+	return apis
+}
+
+/////////////////////////////////////////////////////////////////////
+// SECTION: p2p.protocol interface
+/////////////////////////////////////////////////////////////////////
+func (swap *SwapProtocol) Run(peer *p2p.Peer, rw p2p.MsgReadWriter) error {
+	p := protocols.NewPeer(peer, rw, swapSpec)
+	sp := NewPeer(p, swap)
+	swap.setPeer(sp)
+	defer swap.deletePeer(sp)
+	defer swap.Close()
+	return sp.Run(sp.handleSwapMsg)
+}
+
+func (swap *SwapProtocol) NodeInfo() interface{} {
+	return nil
+}
+
+func (swap *SwapProtocol) PeerInfo(id discover.NodeID) interface{} {
+	return nil
+}
+
+//------------------------------------------------------------------------------------------
+func (swap *SwapProtocol) Close() {
+}
+
+func (swap *SwapProtocol) getPeer(peerId discover.NodeID) *SwapProtocolPeer {
+	swap.peersMu.RLock()
+	defer swap.peersMu.RUnlock()
+
+	return swap.peers[peerId]
+}
+
+func (swap *SwapProtocol) setPeer(peer *SwapProtocolPeer) {
+	swap.peersMu.Lock()
+	defer swap.peersMu.Unlock()
+
+	swap.peers[peer.ID()] = peer
+	metrics.GetOrRegisterGauge("registry.peers", nil).Update(int64(len(swap.peers)))
+}
+
+func (swap *SwapProtocol) deletePeer(peer *SwapProtocolPeer) {
+	swap.peersMu.Lock()
+	defer swap.peersMu.Unlock()
+
+	delete(swap.peers, peer.ID())
+	metrics.GetOrRegisterGauge("registry.peers", nil).Update(int64(len(swap.peers)))
+	swap.peersMu.Unlock()
+}
+
+func (swap *SwapProtocol) peersCount() (c int) {
+	swap.peersMu.Lock()
+	c = len(swap.peers)
+	swap.peersMu.Unlock()
+	return
+}
+
+func (p *SwapProtocolPeer) handleSwapMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *IssueChequeMsg:
```

Comment:
what are these supposed to be? I dont get what the two message types are

Reply:
My understanding was that we are modeling the protocol along the network paper https://www.sharelatex.com/project/59c298da5b3a6e3b7ec80d56 "Generalized swap swear and swindle games". There, we use the terms "Issue cheque" and "Redeem cheque" in figure 1, chapter 2.1

Reply:
No. 

According to https://www.sharelatex.com/project/59c298da5b3a6e3b7ec80d56, 
"Figure 1: The basic interaction sequence for swap chequebooks", 

when a node A gets too indebted with node B, it sends a Cheque to node B (`issue cheque` --> `IssueChequeMsg`). If then node B wants to cash this cheque in, it sends the `redeem cheque` (`RedeemChequeMsg`) message to node A. 

According to that same paper, these interactions are off-chain, and only handling the `RedeemChequeMsg` triggers a blockchain transaction.